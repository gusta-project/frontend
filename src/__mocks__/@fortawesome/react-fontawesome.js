import React from 'react';
import PropTypes from 'prop-types';

export function FontAwesomeIcon({ icon }) {
  const iconClass = Array.isArray(icon) ? icon.join('-') : icon;

  return <i className={`fa ${iconClass}`} />;
}

FontAwesomeIcon.propTypes = {
  icon: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      prefix: PropTypes.string,
      iconName: PropTypes.string,
      icon: PropTypes.arrayOf(PropTypes.any)
    })
  ]).isRequired
};
