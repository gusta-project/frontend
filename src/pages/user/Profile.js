import { Helmet } from 'react-helmet';
import { Link } from 'react-router-dom';
import React, { Component } from 'react';
import {
  Container,
  Row,
  Col,
  Card,
  CardGroup,
  ButtonGroup,
  Button
} from 'react-bootstrap';

export default class Profile extends Component {
  render() {
    return (
      <Container>
        <Helmet title="Your Profile" />
        <Row className="text-center">
          <Col>
            <p>
              This is what other users see when they look at your profile.
              <br />
              <Link to="/user-settings">Click here to edit your profile</Link>
            </p>
          </Col>
        </Row>
        <Row className="text-center">
          <Col>
            <h1>xXTeddyBearSlayer69Xx</h1>
            <img src="/media/harold.jpeg" alt="Profile pic" />
          </Col>
        </Row>
        <Row className="text-center">
          <Col>
            <ButtonGroup>
              <Button className="button-animation" variant="primary">
                <span>Message</span>
              </Button>
              <Button className="button-animation" variant="primary">
                <span>Follow</span>
              </Button>
              <Button className="button-animation" variant="primary">
                <span>Report</span>
              </Button>
            </ButtonGroup>
            <hr />
          </Col>
        </Row>
        <Row className="text-center">
          <Col>
            <h2>Newest recipes</h2>
            <CardGroup>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-1.jpg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">Strawberry Milkshake</Link>
                  </Card.Title>
                  <Card.Text>created 10/13/2018</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="Checkmark emoji">
                      ✔
                    </span>
                    ️ You have everything!
                  </small>
                </Card.Footer>
              </Card>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-2.jpg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">Bomb Pop</Link>
                  </Card.Title>
                  <Card.Text>created 12/1/2018</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="Checkmark emoji">
                      ✔
                    </span>
                    ️ You have everything!
                  </small>
                </Card.Footer>
              </Card>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-3.jpg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">Tiramisu</Link>
                  </Card.Title>
                  <Card.Text>created 6/28/2019</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="X emoji">
                      ❌
                    </span>{' '}
                    You are missing 1 flavor
                  </small>
                </Card.Footer>
              </Card>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-4.jpeg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">RY4 Tobaccy</Link>
                  </Card.Title>
                  <Card.Text>created 4/4/2019</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="Checkmark emoji">
                      ✔
                    </span>
                    ️ You have everything!
                  </small>
                </Card.Footer>
              </Card>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-5.jpeg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">Fruit Punch</Link>
                  </Card.Title>
                  <Card.Text>created 1/1/2018</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="X emoji">
                      ❌
                    </span>{' '}
                    You are missing 3 flavors
                  </small>
                </Card.Footer>
              </Card>
            </CardGroup>
            <hr />
          </Col>
        </Row>
        <Row className="text-center">
          <Col>
            <h2>Most popular recipes</h2>
            <CardGroup>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-1.jpg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">Strawberry Milkshake</Link>
                  </Card.Title>
                  <Card.Text>created 10/13/2018</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="Checkmark emoji">
                      ✔
                    </span>
                    ️ You have everything!
                  </small>
                </Card.Footer>
              </Card>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-2.jpg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">Bomb Pop</Link>
                  </Card.Title>
                  <Card.Text>created 12/1/2018</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="Checkmark emoji">
                      ✔
                    </span>
                    ️ You have everything!
                  </small>
                </Card.Footer>
              </Card>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-3.jpg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">Tiramisu</Link>
                  </Card.Title>
                  <Card.Text>created 6/28/2019</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="X emoji">
                      ❌
                    </span>{' '}
                    You are missing 1 flavor
                  </small>
                </Card.Footer>
              </Card>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-4.jpeg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">RY4 Tobaccy</Link>
                  </Card.Title>
                  <Card.Text>created 4/4/2019</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="Checkmark emoji">
                      ✔
                    </span>
                    ️ You have everything!
                  </small>
                </Card.Footer>
              </Card>
              <Card>
                <Card.Img
                  variant="top"
                  src="/media/card-test-5.jpeg"
                  className="img-fluid w-75 mx-auto"
                />
                <Card.Body>
                  <Card.Title>
                    <Link to="#">Fruit Punch</Link>
                  </Card.Title>
                  <Card.Text>created 1/1/2018</Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small>
                    <span role="img" aria-label="X emoji">
                      ❌
                    </span>{' '}
                    You are missing 3 flavors
                  </small>
                </Card.Footer>
              </Card>
            </CardGroup>
          </Col>
        </Row>
      </Container>
    );
  }
}
