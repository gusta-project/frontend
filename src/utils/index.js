import dayjs from 'dayjs';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = (endpoint) => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

/**
 * Gets the initial Redux state from localStorage and the source code defaults.
 */
export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
