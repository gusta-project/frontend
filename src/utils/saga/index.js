import counterUtilities from './counter';
import pageUtilities from './pager';
import toastUtilities from './toasts';

export default {
  ...counterUtilities,
  ...pageUtilities,
  ...toastUtilities
};
