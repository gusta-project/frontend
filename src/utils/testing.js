import React, { createElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

/**
 * Creates a mock component which takes a set of props specified at runtime.
 *
 * @param {string} name The component/element name (in PascalCase)
 * @param {object} props An object containing props to setup
 * @return {object} Mock component with specified name and props
 */
export const mockComponent = (name, props = {}) => () =>
  createElement(name, props, props.children);

/**
 * Creates a mock component which will expose its props for snapshot testing purposes
 * @param {*} name The component/element name (in PascalCase)
 */
export const mockComponentWithProps = (name) => (props) =>
  createElement(name, props, props.children);

/**
 * Wraps a React component in a <MemoryRouter> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withMemoryRouter = (WrappedComponent, routerProps = {}) => (
  props
) => (
  <MemoryRouter {...routerProps}>
    <WrappedComponent {...props} />
  </MemoryRouter>
);

/**
 * Wraps a React component in a <Provider> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 * @param {object} store Mock store
 */
export const withProvider = (WrappedComponent, store) => (props) => (
  <Provider store={store}>
    <WrappedComponent {...props} />
  </Provider>
);
