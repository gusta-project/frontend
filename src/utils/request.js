import axios from 'axios';
import { buildUrl } from 'utils';
import { call, race, delay } from 'redux-saga/effects';

/**
 * Make an XMLHttpRequest to the specified endpoint with the specified request data.
 *
 * @param {object} endpoint An object with `url` and `method` keys
 * @param {object} data An object to use as the request body
 */
class Request {
  *execute({ endpoint, data, options = {} }) {
    try {
      // ensure the endpoint was supplied
      if (!endpoint) {
        throw new Error('No endpoint provided!');
      }

      // ensure properties required by axios are present
      const { url, method } = endpoint;

      if (!url) {
        throw new Error('Endpoint is missing URL!');
      }

      if (!method) {
        throw new Error('Endpoint is missing method!');
      }

      // default timeout of 30 seconds
      const { timeout = 30000 } = options;
      const requestUrl = yield call(buildUrl, endpoint);

      // start a race between the request and a timer, cancel the loser
      const { response } = yield race({
        response: call(axios, {
          url: requestUrl,
          method,
          data
        }),
        timeout: delay(timeout)
      });

      if (!response) {
        throw new Error('Request timed out!');
      }

      const { status } = response;
      const success = status >= 200 && status < 400;

      if (success) {
        return {
          success,
          response
        };
      } else {
        return {
          success,
          error: response
        };
      }
    } catch (error) {
      return {
        success: false,
        error
      };
    }
  }
}

const instance = new Request();

export default instance;
