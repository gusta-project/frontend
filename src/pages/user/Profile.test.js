import React from 'react';
import renderer from 'react-test-renderer';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Profile } from './Profile';
import { withMemoryRouter } from 'utils';

describe('<Profile />', () => {
  const match = { params: { id: 1 }, path: '/user/profile' };
  const collection = {
    name: 'mixn',
    userId: 10,
    location: 'WhyNotMinot, CAN',
    bio: 'FA Cheez Whiz is my favorite.',
    url: 'itsfreezing.com',
    gravatar: 'xxx'
  };
  const actions = {
    requestCurrentUserProfile: jest.fn(),
    requestProfile: jest.fn(),
    updateProfile: jest.fn()
  };

  const RoutedProfile = withMemoryRouter(Profile);

  const props = {
    match,
    collection,
    actions,
    editing: false,
    loggedIn: true,
    currentUser: true
  };

  const component = renderer.create(<RoutedProfile {...props} />);

  const { instance } = component.root.findByType(Profile);

  it('renders visited user profile correctly', () => {
    const uprops = {
      ...props,
      match: {
        params: { id: 1, userName: 'david' },
        path: '/user/david'
      },
      currentUser: false
    };
    const tree = renderer.create(<RoutedProfile {...uprops} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders current user profile correctly', () => {
    const tree = renderer.create(<RoutedProfile {...props} />).toJSON;

    expect(tree).toMatchSnapshot();
  });

  it('can get gravatar', () => {
    expect(instance).toBeDefined();
    expect(instance.gravatar).toEqual(
      'https://www.gravatar.com/avatar/xxx?s=250'
    );
  });

  it('can get website', () => {
    expect(instance).toBeDefined();
    expect(instance.website).toEqual(
      <a href="//itsfreezing.com">itsfreezing.com</a>
    );
  });

  it('can get locationIcon', () => {
    expect(instance).toBeDefined();
    expect(instance.locationIcon).toEqual(<FontAwesomeIcon icon="city" />);
  });

  it('can get webIcon', () => {
    expect(instance).toBeDefined();
    expect(instance.webIcon).toEqual(<FontAwesomeIcon icon="globe" />);
  });

  it('can get bioIcon', () => {
    expect(instance).toBeDefined();
    expect(instance.bioIcon).toEqual(<FontAwesomeIcon icon="info-circle" />);
  });

  it('can handleEditor show', () => {
    expect(instance).toBeDefined();
    instance.handleEditor(true);
    expect(instance.state.editing).toEqual(true);
  });

  it('can handleEditor hide', () => {
    expect(instance).toBeDefined();
    instance.handleEditor(false);
    expect(instance.state.editing).toEqual(false);
  });

  it('can handleSubmit', () => {
    expect(instance).toBeDefined();
    const profile = {
      location: 'WhyNotMinot, CAN',
      bio: 'FA Cheez Whiz is my favorite.',
      url: 'itsfreezing.com'
    };

    instance.handleSubmit(profile);
    expect(actions.updateProfile).toBeCalledWith(profile);
    expect(instance.state.editing).toEqual(false);
  });

  it('can display profileEditor', () => {
    expect(instance.profileEditor).not.toBeNull();
  });
});
