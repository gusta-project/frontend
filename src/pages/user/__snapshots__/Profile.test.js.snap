// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> renders current user profile correctly 1`] = `[Function]`;

exports[`<Profile /> renders visited user profile correctly 1`] = `
<div
  className="container"
>
  <div
    className="text-center row"
  >
    <div
      className="col"
    >
      <p>
        This is what other users see when they look at your profile.
        <br />
        <a
          href="/user-settings"
          onClick={[Function]}
        >
          Click here to edit your profile
        </a>
      </p>
    </div>
  </div>
  <div
    className="text-center mb-3 row"
  >
    <div
      className="col-md-4"
    >
      <div
        className="card"
      >
        <img
          className="card-img-top"
          src="https://www.gravatar.com/avatar/xxx?s=250"
        />
        <h1
          className="card-header"
        >
          mixn
        </h1>
        <div
          className="card-body"
        >
          <div>
            <i
              className="fa city"
            />
             
            WhyNotMinot, CAN
          </div>
          <div>
            <i
              className="fa globe"
            />
             
            <a
              href="//itsfreezing.com"
            >
              itsfreezing.com
            </a>
          </div>
        </div>
      </div>
      <div
        className="text-left card"
      >
        <div
          className="card-body"
        >
          FA Cheez Whiz is my favorite.
        </div>
      </div>
      <div
        className="mt-3 btn-group"
        role="group"
      >
        <button
          className="button-animation btn btn-primary"
          disabled={false}
          type="button"
        >
          <span>
            Message
          </span>
        </button>
        <button
          className="button-animation btn btn-primary"
          disabled={false}
          type="button"
        >
          <span>
            Follow
          </span>
        </button>
        <button
          className="button-animation btn btn-primary"
          disabled={false}
          type="button"
        >
          <span>
            Report
          </span>
        </button>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="text-center row"
      >
        <div
          className="col"
        >
          <h2>
            Coming soon! 
            <sup>
              â„¢
            </sup>
          </h2>
          <hr />
        </div>
      </div>
    </div>
    <hr />
  </div>
</div>
`;
